rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md

```diff
@@ -0,0 +1,542 @@
+        dataset,
+        epochs=100,
+        validation_data=validation_data,
+        # Tune whether to use shuffle.
+        shuffle=hp.Boolean("shuffle"),
+        # Tune whether to use sample_weights.
+        sample_weight=sample_weight if hp.Boolean("sample_weight") else None,
+        # The provided callbacks list contains checkpointing and tensorboard.
+        callbacks=callbacks)
+
+tuner = kt.RandomSearch(
+  hypermodel=MyHyperModel(),
+  objective=kt.Objective('val_mae', 'min'),
+  directory='dir',
+  max_trials=3,
+  executions_per_trial=2,
+  overwrite=True,
+  directory="my_dir",
+  project_name="helloworld",
+)
+
+# Any arg passed to `search()` would be passed to `fit()`.
+tuner.search()
+
+# Exporting the best models.
+models = tuner.get_best_models(num_models=2)
+
+# Retraining the model with the second best hyperparameters.
+second_best_hp = tuner.get_best_hyperparameters(num_models=2)[1]
+hypermodel = MyHyperModel()
+model = hypermodel.build(second_best_hp)
+hypermodel.fit(model, hp=second_best_hp, callbacks=[], dataset=new_dataset)
+```
+
+There are three details in this workflow that are worth attention.
+
+* Similar to `Tuner.run_trial()`, the return value of the fit function supports
+  all different formats.
+
+* The user can use built-in callbacks just like in `Tuner.run_trial()`.
+
+* `build()` and `fit()` can share hyperparameters. In this example,
+  `num_features` is shared between the two functions. In `fit()`, we can use
+  `hp.get()` to obtain the value of a hyperparameter used in `build()`.
+
+* We can easily retrain the model with any hyperparameter value set with
+  `hypermodel.build()` and `hypermodel.fit()`.
+
+With these proposed workflows, the user now has the maximum flexibility. Any
+step in an end-to-end machine learning workflow can be tuned. Moreover, the
+changes needed to tune existing Keras code is minimized.
+
+Here we present HyperModel code examples of three important use cases:
+
+* Text tokenization.
+
+* Custom training loop.
+
+* Fine tuning with pretrained weights.
+
+#### Text tokenization
+
+```py
+class MyHyperModel(kt.HyperModel):
+  def build(self, hp):
+    inputs = keras.Input(shape=(10,))
+    outputs = layers.Embedding(
+        # max_token is a hyperparameter also used in text vectorization.
+        input_dim=hp.Int("max_tokens", 100, 500, step=100),
+        output_dim=64)(inputs)
+    outputs = layers.LSTM(hp.Int("units", 32, 128, step=32))(outputs)
+    outputs = layers.Dense(1, activation='sigmoid')(outputs)
+    model = keras.Model(inputs, outputs)
+    model.compile(loss='mse')
+    return model
+  
+  def fit(self, hp, model, x, y, validation_data, callbacks, **kwargs):
+    # Create and adapt the text vectorizer.
+    text_vectorizer = layers.TextVectorization(
+        # The max_tokens is a hyperparameter created in build().
+        max_tokens=hp.get("max_tokens"),
+        output_mode="int",
+        output_sequence_length=10)
+    text_vectorizer.adapt(x)
```

Comment:
Recomputing the vocab all the time isn't the best. Maybe we should just load a truncated version of the vocab file.